From ddf409aed9c78d04cc32d12cf4fe86d1adae7e15 Mon Sep 17 00:00:00 2001
From: Ivan Mironov <mironov.ivan@gmail.com>
Date: Tue, 4 Mar 2025 07:28:41 +0100
Subject: [PATCH] Support enabling features by environment variable

This adds new environment variable `ZED_ENABLE_EXPERIMENTAL_FEATURES`
that accepts comma-separated list of features:

	ZED_ENABLE_EXPERIMENTAL_FEATURES=feat1,feat2 zed

Names of all existing features can be found in
`crates/feature_flags/src/feature_flags.rs`. Note that some features may
be broken or require proprietary remote services which you are not
allowed to access.

This patch also removes the ability to disable features at runtime.
---
 crates/feature_flags/src/feature_flags.rs |  4 +++-
 crates/zed/src/main.rs                    | 23 ++++++++++++++++++++++-
 2 files changed, 25 insertions(+), 2 deletions(-)

diff --git a/crates/feature_flags/src/feature_flags.rs b/crates/feature_flags/src/feature_flags.rs
index 1b66b291b527..6408ae6c51e8 100644
--- a/crates/feature_flags/src/feature_flags.rs
+++ b/crates/feature_flags/src/feature_flags.rs
@@ -145,7 +145,9 @@ impl FeatureFlagAppExt for App {
     fn update_flags(&mut self, staff: bool, flags: Vec<String>) {
         let feature_flags = self.default_global::<FeatureFlags>();
         feature_flags.staff = staff;
-        feature_flags.flags = flags;
+        feature_flags.flags.extend(flags);
+        feature_flags.flags.sort();
+        feature_flags.flags.dedup();
     }
 
     fn set_staff(&mut self, staff: bool) {
diff --git a/crates/zed/src/main.rs b/crates/zed/src/main.rs
index 9dba1b427d35..c2340368d69c 100644
--- a/crates/zed/src/main.rs
+++ b/crates/zed/src/main.rs
@@ -12,6 +12,7 @@ use crashes::InitCrashHandler;
 use db::kvp::{GLOBAL_KEY_VALUE_STORE, KEY_VALUE_STORE};
 use editor::Editor;
 use extension::ExtensionHostProxy;
+use feature_flags::FeatureFlagAppExt;
 use fs::{Fs, RealFs};
 use futures::{StreamExt, channel::oneshot, future};
 use git::GitHostingProviderRegistry;
@@ -33,7 +34,7 @@ use release_channel::{AppCommitSha, AppVersion, ReleaseChannel};
 use session::{AppSession, Session};
 use settings::{BaseKeymap, Settings, SettingsStore, watch_config_file};
 use std::{
-    env,
+    env::{self, VarError},
     io::{self, IsTerminal},
     path::{Path, PathBuf},
     process,
@@ -694,6 +695,26 @@ pub fn main() {
         cx.set_menus(menus);
         initialize_workspace(app_state.clone(), prompt_builder, cx);
 
+        let feature_flags = env::var("ZED_ENABLE_EXPERIMENTAL_FEATURES")
+            .and_then(|s| {
+                if s.is_empty() {
+                    Err(VarError::NotPresent)
+                } else {
+                    Ok(s)
+                }
+            })
+            .map_or_else(
+                |e| {
+                    if e == VarError::NotPresent {
+                        Vec::new()
+                    } else {
+                        panic!("Invalid ZED_ENABLE_EXPERIMENTAL_FEATURES: {:?}", e)
+                    }
+                },
+                |s| s.split(',').map(|s| s.to_string()).collect(),
+            );
+        cx.update_flags(false, feature_flags);
+
         cx.activate(true);
 
         cx.spawn({
-- 
2.51.1

